### PR TITLE
Dropdown sources using card can select column

### PR DESCRIPTION
**Pull request by qnkhuat**
Implement milestone 1 of https://github.com/metabase/metabase/issues/26873

Previously in https://github.com/metabase/metabase/pull/27008, we were able to set a card as the source for a parameter and the returned values are the 1st column from the card.

This PR lets UI able to specify a column to be returned by providing a `value_field`.

```diff
{
  "name": "Text starts with",
  "slug": "text_starts_with",
  "id": "bc2a98cf",
  "type": "string/=",
  "source_type": "static-list",
  "source_options": {
      "card_id": 1
+     "value_field": ["field", 34, nil]},
}
```

Note for UI: when the underlining card change (change query, remove column etc..) and BE can't find the corresponding field with the specified field_ref. BE will throw an exception and returns 400. 
Maybe FE should catch this and show a friendly error?

--- 

Also, we change the way we query from a card to using mbql. 
This makes getting values more efficient because we'll only query what we need from DB and filtering is done within DB itself.

Note: current;y binning does not work because it's unclear if field-ref from FE will contain the full binning info.
This PR will throw an error if you use a binning column.
I added a todo in the epic, will revisit this later.

### CODE REVIEW

File: test/metabase/models/params/custom_values_test.clj [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+(defn- find-field-ref
+  [id-or-name result-metadata]
+  (let [selector (if (string? id-or-name)
+                   :name
+                   :id)]
+    (-> (filter #(= id-or-name (selector %)) result-metadata)
+        first
+        :field_ref)))
+
+(deftest with-mbql-card-test
```
Comment (calherries) [inline test/metabase/models/params/custom_values_test.clj:18]: Can we test this works with a field that has a `:binning` strategy, and also a field with `:temporal-unit` bucketing?

File: src/metabase/models/params/custom_values.clj [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      (throw (ex-info (tru "No matching field found")
+                      {:status-code 400
+                       :card-id     card-id
+                       :field-ref   field-ref})))
+    (case (first field-ref)
+      :aggregation
+      [:field (:name field) {:base-type ((some-fn :effective_type :base_type) field)}]
+
+      :field
+      field-ref
```
Comment (calherries) [inline src/metabase/models/params/custom_values.clj:37]: Unless this is done on the frontend, I think you'll have to strip out :binning and :temporal-unit from the field here. Ideally we'd have a universal way to key columns that didn't include this information, and it should probably live in a mbql shared namespace so it can be called outside the query-processor namespaces. 

See this for the existing logic that's used in the query processor
https://github.com/metabase/metabase/blob/464ec9e5bede2aec50391d50f8f7b5b600e61f96/src/metabase/query_processor/middleware/add_implicit_clauses.clj#L58

Comment (calherries) [reply src/metabase/models/params/custom_values.clj:37]: I found this discussion between Cam and Dan that might give more context.
https://github.com/metabase/metabase/pull/19158#r770039976

Note this is an existing issue with `field-ref->key`, so don't look to copy that function.

Comment (qnkhuat) [reply src/metabase/models/params/custom_values.clj:37]: that's a good point.

Given the inherent issue with identifying a key, I wonder if it's better to let UI construct a `field` so that we could directly use it in the mbql. I believe UI already did this because we could select any fields to be returned from a card/table in the Question editor.


This way we don't have to deal with field finding and the `binning, temporal-unit` options

File: shared/src/metabase/mbql/schema.cljc [hunk 3]
```diff
@@ -416,6 +416,10 @@
       validate-temporal-unit
       no-binning-options-at-top-level))
 
+(def field-options-for-identification
+  "Set of FieldOptions that only mattered for idenfiticaiton purposes."
```
Comment (calherries) [inline shared/src/metabase/mbql/schema.cljc:420]: Serious typo here with idenfiticaiton 😆

Comment (qnkhuat) [reply shared/src/metabase/mbql/schema.cljc:420]: 😆 can't identify the typo.

will fix!

File: test/metabase/test.clj [hunk 4]
```diff
@@ -136,7 +136,8 @@
  [qp
   compile
   preprocess
-  process-query]
+  process-query
+  query->expected-cols]
```
Comment (calherries) [inline test/metabase/test.clj:140]: I'm curious, why aren't you using `qp.test-util/card-with-source-metadata-for-query`? I think the point of that function is to prepare card's `result_metadata`.

Comment (qnkhuat) [reply test/metabase/test.clj:140]: no reasons tbh, I just didn't know about that function.

looks like a neat one. Will replace to use it.

Comment (qnkhuat) [reply test/metabase/test.clj]: Changed https://github.com/metabase/metabase/pull/27343/commits/d8766d67ed428a5b0dc56776774e737e58f88daa and https://github.com/metabase/metabase/pull/27343/commits/1e40f7fc3ae564ad14278ff583b638761e992467

File: test/metabase/models/params/custom_values_test.clj (hunk 1 above)
Comment (qnkhuat) [reply test/metabase/models/params/custom_values_test.clj:18]: added tests for temporal-unit bucketing.

With binning, things get a bit weird because I'm not sure if field-ref from FE will contain the binning info or not (the full `[:field 42 {:binning {:strategy ... }}]`

I'll just throw an error for now: https://metaboat.slack.com/archives/C04DAKKDMFT/p1671632554048439?thread_ts=1671611125.666239&cid=C04DAKKDMFT.

Will revisit this later.

Comment (qnkhuat) [reply test/metabase/models/params/custom_values_test.clj:18]: Added an item to the nice to have in this list: https://github.com/metabase/metabase/issues/26873

Comment (qnkhuat) [reply test/metabase/models/params/custom_values_test.clj:18]: added tests for temporal bucketing.

For binning it's not clear to me whether the binning info will be included in the field_ref.

if it's already included then it'll work with no problems.
I added a todo to our milestone to review this later: https://github.com/metabase/metabase/issues/26873

Currently I'll just throw an error if users select binning column.

File: src/metabase/util/schema.clj [hunk 5]
```diff
@@ -288,6 +290,15 @@
...
     (deferred-tru "value must be a valid field semantic or relation type (keyword or string).")))
 
+(def FieldRef
+  "Schema for a valid Field Referenced."
+  (s/cond-pre
+    mbql.s/FieldOrAggregationReference
+    (s/pred
+      (comp (complement (s/checker mbql.s/FieldOrAggregationReference))
+            mbql.normalize/normalize-tokens)
+      "Field or aggregation reference as it comes in to the API")))
```
Comment (calherries) [inline src/metabase/util/schema.clj:300]: Why do we need this schema? Because the tokens are normalized on the BE, but not the FE?

File: src/metabase/models/params/custom_values.clj [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+     :query    (merge
+                 {:source-table (format "card__%d" card-id)
+                  :fields       [value-field]
+                  :limit        *max-rows*}
+                 (when query
+                   {:filter [:contains [:lower value-field] (str/lower-case query)]}))
+     :middleware {:disable-remaps? true}}))
+
+(s/defn values-from-card
+  "Get a column from of a card using field_ref.
```
Comment (calherries) [inline src/metabase/models/params/custom_values.clj:69]: Typo

File: src/metabase/api/dashboard.clj [hunk 7]
```diff
@@ -738,10 +738,14 @@
   [{source-options :source_options :as _param} query]
   (when-let [values (:values source-options)]
     {:values (if query
-               (parameter-card/query-matches query values)
+               (custom-values/query-matches query values)
```
Comment (calherries) [inline src/metabase/api/dashboard.clj:741]: If `query` is nil then `query-matches` will return `values` anyway. Either remove the branching from here, or in the body of `query-matches`

Comment (calherries) [reply src/metabase/api/dashboard.clj:741]: Oh, I just realised this problem was here before, you just moved query-matches. Even so, best tidy that up now

File: src/metabase/api/dashboard.clj [hunk 8]
```diff
@@ -764,7 +768,7 @@
                         :status-code     400})))
      (case (:source_type param)
        "static-list" (static-parameter-values param query)
-       "card"        (parameter-card/values-for-dashboard dashboard param-key query)
+       "card"        (card-parameter-values param query)
        (chain-filter dashboard param-key constraint-param-key->value query)))))
```
Comment (calherries) [inline src/metabase/api/dashboard.clj:772]: Also something from before, but this should explicitly handle `nil`

File: test/metabase/models/params/custom_values_test.clj (hunk 1 above)
Comment (calherries) [reply test/metabase/models/params/custom_values_test.clj:18]: When you do revisit it, I still think temporal-unit should be explicitly absent from the MBQL query too.

The test passes because applying temporal-unit on a field in the MBQL query returns the same results as querying the field with no temporal-unit. But it's only because the field in the source card's query has had temporal bucketing applied previously, and temporal bucketing is idempotent. It smells a bit magic. Ideally be doing is coming up with a consistent way to query fields / aggregations without applying temporal bucketing or binning or _anything_ like that.

File: src/metabase/models/params/custom_values.clj [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+              (throw (ex-info (tru "Binning column not supported")
+                              {:status-code 400
+                               :card-id     card-id
+                               :field-ref   field-ref})))
+        [:field identifier (select-keys options mbql.s/field-options-for-identification)])
+
+      (throw (ex-info (tru "Invalid field-ref type. Must be a field or aggregation.")
+                      {:status-code 400
+                       :card-id     card-id
+                       :field-ref   field-ref})))))
```
Comment (calherries) [inline src/metabase/models/params/custom_values.clj:51]: Throwing exceptions with `:status-code` couples this function with api endpoints. If code is in a `metabase.models` namespace I would have assumed it should be decoupled from the api. So it should either be impossible to reach this code path (in that case this exception doesn't need to be thrown), or it shouldn't live in the `metabase.models` namespace. The exception for binning seems fine as long as we clean it up later. btw I think this is an existing issue with `metabase.models.params.chain-filter` too.

**Review by calherries — CHANGES_REQUESTED** [review]

File: src/metabase/models/params/custom_values.clj [hunk 10]
```diff
@@ -0,0 +1,110 @@
+(ns metabase.models.params.custom-values
```
[the remark this answers is not in the document]
Comment (calherries) [reply src/metabase/models/params/custom_values.clj:1]: I see it's to differentiate it from `field-values`, which is the default type of parameter values source. Perhaps we should avoid creating a name for something that doesn't really have meaning (by that I mean, there's no logical similarity between source=card and source=list other than the fact that it's not source=field). I'd suggest separating these blocks into separate namespaces.

[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/models/params/custom_values.clj:1]: that's a good point. I name it custom-values solely because the project name is custom-values.
Didn't give it much thought.

I'll move this into `metabase.models.params.field-values` 👍

File: src/metabase/models/params/custom_values.clj [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+  =>
+  {:values          [\"Red Medicine\"]
+   :has_more_values false}
+  "
+  ([card-id value-field-ref]
+   (values-from-card card-id value-field-ref nil))
+
+  ([card-id         :- su/IntGreaterThanZero
+    value-field-ref :- su/FieldRef
+    query           :- (s/maybe su/NonBlankString)]
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/models/params/custom_values.clj:86]: For consistency, I think we should use `query` altogether for now.

As you pointed out we're using `query` for all endpoints that get param-values (dashboard, embedding, public).

If we want to we could rename all of them to something clearer like `search-term`.

File: src/metabase/models/params/custom_values.clj (hunk 9 above)
Comment (qnkhuat) [reply src/metabase/models/params/custom_values.clj:51]: I hear what you're saying but I think we don't have strict rules on whether the exception for API  should only be placed in the API namespace.

If we want to do that then pretty much any call to `models` namespace will need a try-catch and proper exception naming so API namespace could selectively catch what is intentionally thrown.

FWIW a lot of models namespace throw status-code too: permissions, collection, card to name a field.
I'm saying this not to mean it's okay to do so, but unless we have a better error-handling mechanism I think it's okay to throw them in models ns for now.


On whether this code path is reached or not: yes, I didn't have a test for it but I have a test for the `no matching field found` case: https://github.com/metabase/metabase/blob/afb4eb38b7afd5f96dfcbd8a31556535c343c58c/test/metabase/api/dashboard_test.clj#L2116

File: src/metabase/util/schema.clj (hunk 5 above)
Comment (qnkhuat) [reply src/metabase/util/schema.clj:300]: Ummm, I'm not sure. I just moved this over from `src/metabase/sync/analyze/query_results.clj`

File: test/metabase/models/params/custom_values_test.clj (hunk 1 above)
Comment (qnkhuat) [reply test/metabase/models/params/custom_values_test.clj:18]: Isn't the temporal-unit already excluded?
Currently only `:source-field :join-alias :base-type` is selected. so the temporal-unit returned by the card will be what it set in the original card.

File: src/metabase/api/dashboard.clj [hunk 12]
```diff
@@ -738,10 +738,14 @@
   [{source-options :source_options :as _param} query]
   (when-let [values (:values source-options)]
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/api/dashboard.clj:739]: yep. Good catch!

File: src/metabase/api/dashboard.clj (hunk 7 above)
Comment (qnkhuat) [reply src/metabase/api/dashboard.clj:741]: I removed the branching in `query-matches`. having `query-matches` handle the branching seems to be a bit of magic if you're reading code from this namespace.

File: src/metabase/models/params/custom_values.clj (hunk 11 above)
[the remark this answers is not in the document]
Comment (calherries) [reply src/metabase/models/params/custom_values.clj:86]: Cool, yeah, I would leave that for a separate PR

File: src/metabase/models/params/custom_values.clj [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+                               :card-id     card-id
+                               :field-ref   field-ref})))
+        [:field identifier (select-keys options mbql.s/field-options-for-identification)])
+
+      (throw (ex-info (tru "Invalid field-ref type. Must be a field or aggregation.")
+                      {:status-code 400
+                       :card-id     card-id
+                       :field-ref   field-ref})))))
+
+(defn- custom-values-query
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/models/params/custom_values.clj:53]: good point.

File: src/metabase/util/schema.clj (hunk 5 above)
Comment (calherries) [reply src/metabase/util/schema.clj:300]: It says why in this comment 
https://github.com/metabase/metabase/blob/769bba1ade874c373eef23dc6162edd0f2f4bbcf/src/metabase/sync/analyze/query_results.clj#L28

In uses of `ParameterSourceOptions`, are the tokens normalized already, or not? I'm not sure you need both options as in `src/metabase/sync/analyze/query_results.clj`

File: test/metabase/models/params/custom_values_test.clj (hunk 1 above)
Comment (calherries) [reply test/metabase/models/params/custom_values_test.clj]: Oh, I didn't realise you were only selecting those. I thought you were throwing an error on binning because you were defferring how to deal with those extra field options. If you're already selecting only :source-field, :join-alias, and :base-type, then isn't binning already dealt with and you don't need to throw an exception? It seems like you're already handling binning fine.

Comment (qnkhuat) [reply test/metabase/models/params/custom_values_test.clj]: When I try from the UI, it includes a `binning` map.

Even though I tried to call the same query but remove the `binning` map and it returns the sam result.

But I'm not sure how important the binning map is. So let just ignore it for now and review this later.
Screenshot 2022-12-22 at 16 17 37

Comment (calherries) [reply test/metabase/models/params/custom_values_test.clj]: Oh, of course. It makes sense to me now. Binning is idempotent too, like temporal unit. If you bin a column into 10 bins, and bin that again, and again etc it will return the same result as just binning once. So binning and temporal-unit don't need to be stripped out to produce the correct results. But it smells. The MBQL query shouldn't have any options that apply additional operations other than just selecting the field.

File: src/metabase/api/dashboard.clj (hunk 7 above)
Comment (calherries) [reply src/metabase/api/dashboard.clj]: You removed the branching already? I don't see it yet

File: src/metabase/util/schema.clj [hunk 14]
```diff
@@ -364,8 +375,14 @@
...
-  {(s/optional-key :values) (s/cond-pre [s/Any])
-   (s/optional-key :card_id) IntGreaterThanZero})
+  {;; for source_type = 'static-list'
+   (s/optional-key :values)          (s/cond-pre [s/Any])
+
+   ;; for source_type = 'card'
+   (s/optional-key :card_id)         IntGreaterThanZero
+   (s/optional-key :value_field_ref) FieldRef
+   ;; label_field_ref is optional
+   (s/optional-key :label_field_ref) FieldRef})
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/util/schema.clj:383]: yea it should. but I'm having a branch depend on this so lemme have this merge and the rename branch merge. Then I'll do that. otherwise it'll be conflict hell :'(

File: src/metabase/models/params/custom_values.clj (hunk 13 above)
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/models/params/custom_values.clj]: renamed in https://github.com/metabase/metabase/pull/27343/commits/373301925d5902d51d0ab1f5e9f2b8f55e496f08

File: src/metabase/models/params/custom_values.clj (hunk 10 above)
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/models/params/custom_values.clj]: fair point. I named it custom-values because it's the project name. admittedly it was not given enough thought.

I created a ns `card-values` and move the code there instead. since `static-list` code is only the `query-matches` function, I'll move it to the api.dashboard for now.

if it grows we will have a dedicated ns for it.

https://github.com/metabase/metabase/pull/27343/commits/373301925d5902d51d0ab1f5e9f2b8f55e496f08

File: src/metabase/api/dashboard.clj (hunk 8 above)
Comment (qnkhuat) [reply src/metabase/api/dashboard.clj]: thanks. handled in https://github.com/metabase/metabase/pull/27343/commits/373301925d5902d51d0ab1f5e9f2b8f55e496f08

File: src/metabase/util/schema.clj (hunk 5 above)
Comment (qnkhuat) [reply src/metabase/util/schema.clj]: you're right. I revert the move and create a new on with with only the checking part for API usage.
https://github.com/metabase/metabase/pull/27343/commits/3e1be713bd80bd34fa83088f0682430ea401e35b

File: src/metabase/api/dashboard.clj (hunk 7 above)
Comment (qnkhuat) [reply src/metabase/api/dashboard.clj]: I removed the branch inside `query-matches` function, not here. Also moved the file to this ns.
https://github.com/metabase/metabase/pull/27343/commits/373301925d5902d51d0ab1f5e9f2b8f55e496f08

File: src/metabase/models/params/custom_values.clj (hunk 9 above)
Comment (qnkhuat) [reply src/metabase/models/params/custom_values.clj]: ah, you're right it's not reachable for that exception :+1. Will remove that throw.

On throw status-code from model ns. I think it's a general problem.
My idea of how we should throw this is we need to have custom exception so that the API can catches only thing it meant to catch( in this case no matching field found so it knows it should returns 400)

Otherwise we'll have to have API catch generic exception `clojure.lang.ExceptionInfo` which will catches all exceptions and just return a generic error. not to mention it doesn't know if it should be a 400 or 404?

to sum up, I agree throwing status-code in model ns is not ideal, but it's not the end of the world given how we're currently doing it across codebase.

Comment (calherries) [reply src/metabase/models/params/custom_values.clj]: I'd still argue 400 errors should be caught at the API layer. In this case the binning exception is okay since that's meant to be temporary and it's not worth fixing in the short term.

File: test/metabase/models/params/custom_values_test.clj (hunk 1 above)
Comment (calherries) [reply test/metabase/models/params/custom_values_test.clj]: Sure, that wouldn't be idempotent. What I meant to say is not "binning is idempotent" in general, but "binning is idempotent when the binning parameters are the same", i.e.
image

Same goes for temporal bucketing.

File: src/metabase/models/params/custom_values.clj (hunk 2 above)
Comment (qnkhuat) [reply src/metabase/models/params/custom_values.clj]: good news, we have finally got around and let UI construct the field 🎉

File: shared/src/metabase/mbql/schema.cljc [hunk 15]
```diff
@@ -416,6 +416,11 @@
       validate-temporal-unit
       no-binning-options-at-top-level))
 
+(def field-options-for-identification
```
Comment (qnkhuat) [inline shared/src/metabase/mbql/schema.cljc:419]: this is not used for this PR, but maybe we just keep it for document purposes?

**Comment by qnkhuat** [comment]
we changed how we present the field in `values_source_config`. previously we use `field-ref` but this approach requires custom work in BE to figure out the exact field to select from a card.

We switched to storing an actual field selection provided by UI. This makes BE much simpler.

File: shared/src/metabase/mbql/schema.cljc (hunk 15 above)
Comment (calherries) [reply shared/src/metabase/mbql/schema.cljc:419]: I'd remove it now and recreate it when we need it. Which will probably be soon given our models foreign key project

File: src/metabase/models/params/card_values.clj [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+  (let [card        (db/select-one Card :id card-id)]
+    {:database (:database_id card)
+     :type     :query
+     :query    (merge
+                 {:source-table (format "card__%d" card-id)
+                  :fields       [value-field]
+                  :limit        *max-rows*}
+                 (when query
+                   {:filter [:contains [:lower value-field] (str/lower-case query)]}))
+     :middleware {:disable-remaps? true}}))
```
Comment (calherries) [inline src/metabase/models/params/card_values.clj:26]: Last question - why is `:disable-remaps?` necessary? I don't see a test for it.

File: src/metabase/models/params/card_values.clj [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    [clojure.string :as str]
+    [metabase.models :refer [Card]]
+    [metabase.query-processor :as qp]
+    [metabase.util.schema :as su]
+    [schema.core :as s]
+    [toucan.db :as db]))
+
+(def ^:dynamic *max-rows*
+  "Maximum number of rows returned when running a card."
+  2000)
```
Comment (calherries) [inline src/metabase/models/params/card_values.clj:13]: Still would appreciate a comment explaining why 2000 was picked. Even if there's no specific reason, leave that as a comment. Otherwise our future selves will be puzzled by this magic number

File: src/metabase/models/params/card_values.clj [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+  ;;  :limit        *max-rows*}
+  =>
+  {:values          [\"Red Medicine\"]
+   :has_more_values false}
+  "
+  ([card-id value-field-ref]
+   (values-from-card card-id value-field-ref nil))
+
+  ([card-id     :- su/IntGreaterThanZero
+    value-field :- su/Field
```
Comment (calherries) [inline src/metabase/models/params/card_values.clj:45]: The schema for `Field` is too loose. I would validate here that the Field is only what is necessary to select a field, and excludes binning etc

**Review by calherries — APPROVED** [review]

**Comment by calherries** [comment]
Looks good to me, I only have a couple more questions but I don't think they're crucial to getting this merged.

**Comment by dpsutton** [comment]
I see that `values-from-card-query` allows you to search with `{:filter [:contains [:lower value-field] (str/lower-case query)]}`. This is an implicit assumption that we are only dealing with string-valued fields right? But lots of tests are testing against floats like `[:field "NAME" {:base-type :type/Float}]` This seems like it won't work.

**Review by dpsutton — APPROVED** [review]
I'm approving but I think we need to assert that the `value-field` is a text type field. We can use our type hierarchy `(metabase.types/field-is-type? :type/Text {:base_type :type/Text})` and call it on the column from result_metadata that we identify with the `value-field`. 

I think this is important because we are starting with search from the beginning so we need to handle the different search options on the different types. And right now we are just using `contains` so we have to stay with text columns until we're ready to venture out.

File: src/metabase/models/params/card_values.clj (hunk 16 above)
Comment (qnkhuat) [reply src/metabase/models/params/card_values.clj:26]: https://www.metabase.com/docs/latest/data-modeling/metadata-editing.html#remapping-column-values

We have a feature to remap fk to a field.
This is done in `add-dimension-projections` middleware.

For values for parameter, we want to disable this feature because of cases where you select an FK field as a values source.
And if remapping is enabled the returned value will not be the `fk` field but the field it's mapped to.

Anw, good catch I'll add a test for it.

File: src/metabase/models/params/card_values.clj (hunk 18 above)
Comment (qnkhuat) [reply src/metabase/models/params/card_values.clj:45]: We're using the same schema on MBQL. And this field is used for the same purpose of building the `:fields` in MBQL, so I don't think we need to be to strict about it.

https://github.com/metabase/metabase/blob/master/shared/src/metabase/mbql/schema.cljc#L1370

File: src/metabase/models/params/card_values.clj (hunk 16 above)
Comment (qnkhuat) [reply src/metabase/models/params/card_values.clj:28]: added here https://github.com/metabase/metabase/pull/27343/commits/939983670dda1fdd1b01056db144ea6257a246af

File: src/metabase/models/params/card_values.clj (hunk 17 above)
Comment (qnkhuat) [reply src/metabase/models/params/card_values.clj:15]: updated : https://github.com/metabase/metabase/pull/27343/commits/939983670dda1fdd1b01056db144ea6257a246af

**Comment by qnkhuat** [comment]
@dpsutton I planned to review the type and add support for numbers in [milestone 2](https://github.com/metabase/metabase/issues/26873).

also seems like identifying a field in result_metadta given a field selector is not a simple matter, I'd like to do it on the next PR.

File: src/metabase/models/params/card_values.clj (hunk 18 above)
Comment (calherries) [reply src/metabase/models/params/card_values.clj:47]: It would be nice to know that value-field doesn't contain any options like `:temporal-unit`,  `:source-field`, or `:join-alias`. It means that it can be safely used in `values-from-card-query` like you're doing without running into issues.

File: src/metabase/models/params/card_values.clj (hunk 16 above)
Comment (calherries) [reply src/metabase/models/params/card_values.clj:28]: > Actually, if we want to go far we could still display the remapped values and swap it with real values when applying in a filter. But that's too much effort for a niche feature IMO.

Yeah I think that's fair, and the use case is exactly what the labels should cover. Thanks
